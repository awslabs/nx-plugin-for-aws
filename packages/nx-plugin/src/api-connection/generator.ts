/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { joinPathFragments, ProjectConfiguration, Tree } from '@nx/devkit';
import { ApiConnectionGeneratorSchema } from './schema';
import trpcReactGenerator from '../trpc/react/generator';
import { hasExportDeclaration } from '../utils/ast';
import { readToml } from '../utils/toml';
import fastApiReactGenerator from '../py/fast-api/react/generator';
import { readProjectConfigurationUnqualified } from '../utils/nx';

/**
 * List of supported source and target project types for api connections
 */
const SUPPORTED_PROJECT_TYPES = [
  'ts#trpc-api',
  'py#fast-api',
  'react',
] as const;

type ProjectType = (typeof SUPPORTED_PROJECT_TYPES)[number];

type Connection = { source: ProjectType; target: ProjectType };

/**
 * Enumerates the supported project connections
 */
const SUPPORTED_CONNECTIONS = [
  { source: 'react', target: 'ts#trpc-api' },
  { source: 'react', target: 'py#fast-api' },
] satisfies Connection[];

type ConnectionKey = (typeof SUPPORTED_CONNECTIONS)[number] extends infer C
  ? C extends Connection
    ? `${C['source']} -> ${C['target']}`
    : never
  : never;

/**
 * Generators for each connection type
 */
const CONNECTION_GENERATORS = {
  'react -> ts#trpc-api': (tree, options) =>
    trpcReactGenerator(tree, {
      frontendProjectName: options.sourceProject,
      backendProjectName: options.targetProject,
    }),
  'react -> py#fast-api': (tree, options) =>
    fastApiReactGenerator(tree, {
      frontendProjectName: options.sourceProject,
      fastApiProjectName: options.targetProject,
    }),
} satisfies Record<
  ConnectionKey,
  (tree: Tree, options: ApiConnectionGeneratorSchema) => Promise<any>
>;

/**
 * Generator for a connection from a project to an API project
 */
export const apiConnectionGenerator = async (
  tree: Tree,
  options: ApiConnectionGeneratorSchema,
) => {
  const sourceType = determineProjectType(tree, options.sourceProject);
  const targetType = determineProjectType(tree, options.targetProject);

  if (!sourceType || !SUPPORTED_PROJECT_TYPES.includes(sourceType)) {
    throw new Error(
      `This generator does not support selected source project ${options.sourceProject}`,
    );
  }
  if (!targetType || !SUPPORTED_PROJECT_TYPES.includes(targetType)) {
    throw new Error(
      `This generator does not support selected target project ${options.targetProject}`,
    );
  }

  const connection = SUPPORTED_CONNECTIONS.find(
    (c) => c.source === sourceType && c.target === targetType,
  );

  if (!connection) {
    throw new Error(
      `This generator does not support a connection from ${options.sourceProject} (${sourceType}) to ${options.targetProject} (${targetType})`,
    );
  }

  return await CONNECTION_GENERATORS[
    `${connection.source} -> ${connection.target}`
  ](tree, options);
};

/**
 * Determine whether the given project is of a known project type
 */
export const determineProjectType = (
  tree: Tree,
  projectName: string,
): ProjectType | undefined => {
  const projectConfiguration = readProjectConfigurationUnqualified(
    tree,
    projectName,
  );

  // NB: if adding new checks, ensure these go from most to least specific
  // eg. react website is more specific than typescript project

  if (isTrpcApi(tree, projectConfiguration)) {
    return 'ts#trpc-api';
  }

  if (isFastApi(tree, projectConfiguration)) {
    return 'py#fast-api';
  }

  if (isReact(tree, projectConfiguration)) {
    return 'react';
  }

  return undefined;
};

const sourceRoot = (projectConfiguration: ProjectConfiguration) =>
  projectConfiguration.sourceRoot ??
  joinPathFragments(projectConfiguration.root, 'src');

const isTrpcApi = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // If the project.json says it's a trpc api, there's no need for introspection
  if ((projectConfiguration.metadata as any)?.apiType === 'trpc') {
    return true;
  }

  // Find the src/index.ts
  const indexTs = tree.read(
    joinPathFragments(sourceRoot(projectConfiguration), 'index.ts'),
    'utf-8',
  );
  if (indexTs === null) {
    return false;
  }

  // If the index file exports an AppRouter, it's a trpc api
  if (hasExportDeclaration(indexTs, 'AppRouter')) {
    return true;
  }

  // If there's a src/router.ts or src/lambdas/router.ts which exports an AppRouter, it's a trpc api
  const trpcRouter =
    tree.read(
      joinPathFragments(sourceRoot(projectConfiguration), 'router.ts'),
      'utf-8',
    ) ??
    tree.read(
      joinPathFragments(
        sourceRoot(projectConfiguration),
        'lambdas',
        'router.ts',
      ),
      'utf-8',
    );

  if (trpcRouter === null) {
    return false;
  }

  if (hasExportDeclaration(trpcRouter, 'AppRouter')) {
    return true;
  }

  return false;
};

const isReact = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // if there's a main.tsx, it's a react app
  return tree.exists(
    joinPathFragments(sourceRoot(projectConfiguration), 'main.tsx'),
  );
};

const isFastApi = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // If the project.json says it's a fast api, there's no need for introspection
  if ((projectConfiguration.metadata as any)?.apiType === 'fast-api') {
    return true;
  }

  const pyProjectPath = joinPathFragments(
    projectConfiguration.root,
    'pyproject.toml',
  );
  if (tree.exists(pyProjectPath)) {
    const pyProject = readToml(tree, pyProjectPath);
    if (((pyProject as any)?.project?.dependencies ?? []).includes('fastapi')) {
      return true;
    }
  }

  return false;
};

export default apiConnectionGenerator;
