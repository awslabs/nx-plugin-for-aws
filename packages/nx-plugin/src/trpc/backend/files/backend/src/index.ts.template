export type { AppRouter } from './router.js';
export type { Context } from './init.js';
