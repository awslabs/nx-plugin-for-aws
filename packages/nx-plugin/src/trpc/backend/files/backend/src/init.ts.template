import { initTRPC } from '@trpc/server';
import {
  createErrorPlugin,
  createLoggerPlugin,
  createMetricsPlugin,
  createTracerPlugin,
  IMiddlewareContext,
} from './middleware/index.js';

export type Context = IMiddlewareContext;

export const t = initTRPC.context<Context>().create();

export const publicProcedure = t.procedure
  .unstable_concat(createLoggerPlugin())
  .unstable_concat(createTracerPlugin())
  .unstable_concat(createMetricsPlugin())
  .unstable_concat(createErrorPlugin());
