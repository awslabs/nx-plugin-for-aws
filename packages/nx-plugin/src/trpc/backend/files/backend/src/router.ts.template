import {
  awsLambdaRequestHandler,
  CreateAWSLambdaContextOptions,
} from '@trpc/server/adapters/aws-lambda';
import { echo } from './procedures/echo.js';
import { t } from './init.js';
<%_ const eventType = computeType === 'ServerlessApiGatewayRestApi' ? 'APIGatewayProxyEvent' : 'APIGatewayProxyEventV2WithIAMAuthorizer'; _%>
import type { <%- eventType %> } from 'aws-lambda';

export const router = t.router;

export const appRouter = router({
  echo,
});

export const handler = awsLambdaRequestHandler({
  router: appRouter,
  createContext: (ctx: CreateAWSLambdaContextOptions<<%- eventType %>>) => ctx,
  <%_ if (computeType === 'ServerlessApiGatewayRestApi') { _%>
  responseMeta: () => ({
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': '*',
    },
  }),
  <%_ } _%>
});

export type AppRouter = typeof appRouter;
