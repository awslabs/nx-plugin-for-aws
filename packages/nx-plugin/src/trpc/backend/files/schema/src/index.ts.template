import { z } from 'zod';

export const EchoSchema = z.object({
  result: z.string(),
});

export type IEcho = z.TypeOf<typeof EchoSchema>;
