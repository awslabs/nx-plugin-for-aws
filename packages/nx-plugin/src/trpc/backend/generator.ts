/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  getPackageManagerCommand,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  Tree,
  updateJson,
} from '@nx/devkit';
import { TrpcBackendGeneratorSchema } from './schema';
import kebabCase from 'lodash.kebabcase';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  sharedConstructsGenerator,
} from '../../utils/shared-constructs';
import tsLibGenerator from '../../ts/lib/generator';
import { getNpmScopePrefix, toScopeAlias } from '../../utils/npm-scope';
import { withVersions } from '../../utils/versions';
import { getRelativePathToRoot } from '../../utils/paths';
import { toClassName } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
export async function trpcBackendGenerator(
  tree: Tree,
  options: TrpcBackendGeneratorSchema,
) {
  await sharedConstructsGenerator(tree);
  const apiNamespace = getNpmScopePrefix(tree);
  const apiNameKebabCase = kebabCase(options.apiName);
  const apiNameClassName = toClassName(options.apiName);
  const projectRoot = joinPathFragments(
    options.directory ?? '.',
    apiNameKebabCase,
  );
  const relativePathToProjectRoot = `${joinPathFragments(
    getRelativePathToRoot(tree, `${getNpmScopePrefix(tree)}common-constructs`),
    projectRoot,
  )}`;
  const schemaRoot = joinPathFragments(projectRoot, 'schema');
  const backendRoot = joinPathFragments(projectRoot, 'backend');
  const backendName = `${apiNameKebabCase}-backend`;
  const schemaName = `${apiNameKebabCase}-schema`;
  const backendProjectName = `${apiNamespace}${backendName}`;
  const schemaProjectName = `${apiNamespace}${schemaName}`;
  const enhancedOptions = {
    backendProjectName,
    backendProjectAlias: toScopeAlias(backendProjectName),
    schemaProjectName,
    schemaProjectAlias: toScopeAlias(schemaProjectName),
    apiNameKebabCase,
    apiNameClassName,
    relativePathToProjectRoot,
    pkgMgrCmd: getPackageManagerCommand().exec,
    ...options,
  };
  await tsLibGenerator(tree, {
    name: backendName,
    directory: projectRoot,
    subDirectory: 'backend',
  });
  await tsLibGenerator(tree, {
    name: schemaName,
    directory: projectRoot,
    subDirectory: 'schema',
  });
  if (
    !tree.exists(
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'trpc-apis',
        `${apiNameKebabCase}.ts`,
      ),
    )
  ) {
    generateFiles(
      tree,
      joinPathFragments(
        __dirname,
        'files',
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
      ),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
      enhancedOptions,
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );
    const shouldGenerateCoreTrpcApiConstruct = !tree.exists(
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'core',
        'trpc-api.ts',
      ),
    );
    if (shouldGenerateCoreTrpcApiConstruct) {
      generateFiles(
        tree,
        joinPathFragments(
          __dirname,
          'files',
          SHARED_CONSTRUCTS_DIR,
          'src',
          'core',
        ),
        joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'core'),
        enhancedOptions,
        {
          overwriteStrategy: OverwriteStrategy.KeepExisting,
        },
      );
    }
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'index.ts',
      ),
      './trpc-apis/index.js',
    );
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'trpc-apis',
        'index.ts',
      ),
      `./${apiNameKebabCase}.js`,
    );
    if (shouldGenerateCoreTrpcApiConstruct) {
      addStarExport(
        tree,
        joinPathFragments(
          PACKAGES_DIR,
          SHARED_CONSTRUCTS_DIR,
          'src',
          'core',
          'index.ts',
        ),
        './trpc-api.js',
      );
    }
  }

  updateJson(
    tree,
    joinPathFragments(backendRoot, 'project.json'),
    (config: ProjectConfiguration) => {
      config.metadata = {
        apiName: options.apiName,
      } as unknown;
      config.targets = {
        ...config.targets,
        serve: {
          executor: 'nx:run-commands',
          options: {
            commands: ['tsx src/local-server.ts'],
            cwd: backendRoot,
          },
        },
      };
      return config;
    },
  );

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'backend'),
    backendRoot,
    enhancedOptions,
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'schema'),
    schemaRoot,
    enhancedOptions,
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );
  tree.delete(joinPathFragments(backendRoot, 'src', 'lib'));
  tree.delete(joinPathFragments(schemaRoot, 'src', 'lib'));
  addDependenciesToPackageJson(
    tree,
    withVersions([
      'aws-xray-sdk-core',
      'zod',
      '@aws-lambda-powertools/logger',
      '@aws-lambda-powertools/metrics',
      '@aws-lambda-powertools/tracer',
      '@trpc/server',
    ]),
    withVersions(['@types/aws-lambda', 'tsx']),
  );
  tree.delete(joinPathFragments(backendRoot, 'package.json'));
  tree.delete(joinPathFragments(schemaRoot, 'package.json'));
  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
}
export default trpcBackendGenerator;
