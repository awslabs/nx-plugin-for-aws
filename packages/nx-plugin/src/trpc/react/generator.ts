/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';
import { ReactGeneratorSchema } from './schema';
import {
  JsxSelfClosingElement,
  ObjectLiteralExpression,
  factory,
  ReturnStatement,
  Block,
  NodeFlags,
  JsxTagNameExpression,
  isJsxElement,
  isJsxFragment,
  isJsxSelfClosingElement,
  isJsxChild,
  isJsxExpression,
  isParenthesizedExpression,
} from 'typescript';
import { ast, tsquery } from '@phenomnomnominal/tsquery';
import { runtimeConfigGenerator } from '../../cloudscape-website/runtime-config/generator';
import { toScopeAlias } from '../../utils/npm-scope';
import { withVersions } from '../../utils/versions';
import {
  createJsxElementFromIdentifier,
  replace,
  singleImport,
  destructuredImport,
} from '../../utils/ast';
import { toClassName } from '../../utils/names';
import { formatFilesInSubtree } from '../../utils/format';
export async function reactGenerator(
  tree: Tree,
  options: ReactGeneratorSchema,
) {
  const frontendProjectConfig = readProjectConfiguration(
    tree,
    options.frontendProjectName,
  );
  const backendProjectConfig = readProjectConfiguration(
    tree,
    options.backendProjectName,
  );
  /* eslint-disable @typescript-eslint/no-explicit-any */
  const apiName = (backendProjectConfig.metadata as any)?.apiName;
  const apiNameClassName = toClassName(apiName);
  const backendProjectAlias = toScopeAlias(options.backendProjectName);

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    frontendProjectConfig.root,
    {
      apiName,
      apiNameClassName: toClassName(apiName),
      ...options,
      backendProjectAlias: toScopeAlias(options.backendProjectName),
    },
    {
      overwriteStrategy: OverwriteStrategy.KeepExisting,
    },
  );
  if (options.auth === 'IAM') {
    generateFiles(
      tree,
      joinPathFragments(__dirname, '../../utils/files/website/hooks/sigv4'),
      joinPathFragments(frontendProjectConfig.sourceRoot, 'hooks'),
      {},
    );
  }
  await runtimeConfigGenerator(tree, {
    project: options.frontendProjectName,
  });

  // update main.tsx
  const mainTsxPath = joinPathFragments(
    frontendProjectConfig.sourceRoot,
    'main.tsx',
  );
  singleImport(
    tree,
    mainTsxPath,
    'TrpcClientProviders',
    './components/TrpcClients',
  );
  // Check if TrpcClientProviders already exists
  const mainTsxSource = tree.read(mainTsxPath).toString();
  const mainTsxAst = ast(mainTsxSource);
  const hasProvider =
    tsquery.query(
      mainTsxAst,
      'JsxOpeningElement[tagName.name="TrpcClientProviders"]',
    ).length > 0;
  if (!hasProvider) {
    replace(
      tree,
      mainTsxPath,
      'JsxSelfClosingElement[tagName.name="RouterProvider"]',
      (node: JsxSelfClosingElement) =>
        createJsxElementFromIdentifier('TrpcClientProviders', [node]),
    );
  }
  // update TrpcApis.tsx
  const trpcApisPath = joinPathFragments(
    frontendProjectConfig.sourceRoot,
    'components/TrpcClients/TrpcApis.tsx',
  );
  // Add imports if they don't exist
  destructuredImport(
    tree,
    trpcApisPath,
    ['createIsolatedTrpcClientProvider'],
    './IsolatedTrpcProvider',
  );
  destructuredImport(
    tree,
    trpcApisPath,
    [
      `AppRouter as ${apiNameClassName}AppRouter`,
      `Context as ${apiNameClassName}Context`,
    ],
    backendProjectAlias,
  );
  // Update the export object only if API doesn't exist
  replace(
    tree,
    trpcApisPath,
    'ExportAssignment > ObjectLiteralExpression',
    (node) => {
      const existingProperties = (node as ObjectLiteralExpression).properties;
      if (
        existingProperties.find((p) => p.name?.getText() === apiNameClassName)
      ) {
        return node;
      }
      const newProperty = factory.createPropertyAssignment(
        factory.createIdentifier(apiNameClassName),
        factory.createCallExpression(
          factory.createIdentifier('createIsolatedTrpcClientProvider'),
          [
            factory.createTypeReferenceNode(
              factory.createIdentifier(`${apiNameClassName}AppRouter`),
            ),
            factory.createTypeReferenceNode(
              factory.createIdentifier(`${apiNameClassName}Context`),
            ),
          ],
          [],
        ),
      );
      return factory.createObjectLiteralExpression(
        [...existingProperties, newProperty],
        true,
      );
    },
  );
  // update TrpcClientProviders.tsx
  const trpcClientProvidersPath = joinPathFragments(
    frontendProjectConfig.sourceRoot,
    'components/TrpcClients/TrpcClientProviders.tsx',
  );
  // Add imports
  destructuredImport(
    tree,
    trpcClientProvidersPath,
    ['useRuntimeConfig'],
    '../../hooks/useRuntimeConfig',
  );
  singleImport(tree, trpcClientProvidersPath, 'TrpcApis', './TrpcApis');
  // Check if useContext hook exists and add if it doesn't add it
  const providersSource = tree.read(trpcClientProvidersPath).toString();
  const providersAst = ast(providersSource);
  const hasRuntimeConfig =
    tsquery.query(
      providersAst,
      'VariableDeclaration[name.name="runtimeConfig"] CallExpression[expression.name="useRuntimeConfig"]',
    ).length > 0;
  if (!hasRuntimeConfig) {
    replace(
      tree,
      trpcClientProvidersPath,
      'ArrowFunction > Block',
      (node: Block) => {
        const existingStatements = [...node.statements];
        const runtimeContextVar = factory.createVariableStatement(
          undefined,
          factory.createVariableDeclarationList(
            [
              factory.createVariableDeclaration(
                'runtimeConfig',
                undefined,
                undefined,
                factory.createCallExpression(
                  factory.createIdentifier('useRuntimeConfig'),
                  undefined,
                  [],
                ),
              ),
            ],
            NodeFlags.Const,
          ),
        );
        // Insert the runtimeContext declaration before the return statement
        existingStatements.splice(
          existingStatements.length - 1,
          0,
          runtimeContextVar,
        );
        return factory.createBlock(existingStatements, true);
      },
    );
  }
  // Check if API provider already exists
  const trpcProviderSource = tree.read(trpcClientProvidersPath).toString();
  const trpcProviderAst = ast(trpcProviderSource);
  const hasTrpcProvider =
    tsquery.query(
      trpcProviderAst,
      `JsxOpeningElement PropertyAccessExpression:has(Identifier[name="Provider"]) PropertyAccessExpression:has(Identifier[name="${apiNameClassName}"]) Identifier[name="TrpcApis"]`,
    ).length > 0;
  if (!hasTrpcProvider) {
    // Transform the return statement only if provider doesn't exist
    replace(
      tree,
      trpcClientProvidersPath,
      'ReturnStatement',
      (node: ReturnStatement) => {
        const existingExpression = isParenthesizedExpression(node.expression)
          ? node.expression?.expression
          : node.expression;
        return factory.createReturnStatement(
          factory.createJsxElement(
            factory.createJsxOpeningElement(
              factory.createPropertyAccessExpression(
                factory.createPropertyAccessExpression(
                  factory.createIdentifier('TrpcApis'),
                  factory.createIdentifier(apiNameClassName),
                ),
                factory.createIdentifier('Provider'),
              ) as JsxTagNameExpression,
              undefined,
              factory.createJsxAttributes([
                factory.createJsxAttribute(
                  factory.createIdentifier('apiUrl'),
                  factory.createJsxExpression(
                    undefined,
                    factory.createPropertyAccessExpression(
                      factory.createPropertyAccessExpression(
                        factory.createIdentifier('runtimeConfig'),
                        factory.createIdentifier('httpApis'),
                      ),
                      factory.createIdentifier(apiNameClassName),
                    ),
                  ),
                ),
              ]),
            ),
            [
              isJsxChild(existingExpression) ||
              isJsxElement(existingExpression) ||
              isJsxFragment(existingExpression) ||
              isJsxSelfClosingElement(existingExpression) ||
              isJsxExpression(existingExpression)
                ? existingExpression
                : factory.createJsxExpression(undefined, existingExpression),
            ],
            factory.createJsxClosingElement(
              factory.createPropertyAccessExpression(
                factory.createPropertyAccessExpression(
                  factory.createIdentifier('TrpcApis'),
                  factory.createIdentifier(apiNameClassName),
                ),
                factory.createIdentifier('Provider'),
              ) as JsxTagNameExpression,
            ),
          ),
        );
      },
    );
  }
  addDependenciesToPackageJson(
    tree,
    withVersions([
      '@trpc/client',
      '@trpc/react-query',
      '@tanstack/react-query',
      ...((options.auth === 'IAM'
        ? [
            'oidc-client-ts',
            'aws4fetch',
            '@aws-sdk/client-cognito-identity',
            '@aws-sdk/credential-provider-cognito-identity',
            'react-oidc-context',
          ]
        : []) as any),
    ]),
    withVersions(['@smithy/types']),
  );
  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
}
export default reactGenerator;
