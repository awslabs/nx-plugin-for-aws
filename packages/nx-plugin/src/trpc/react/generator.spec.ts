/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from 'nx/src/devkit-testing-exports';
import { reactGenerator } from './generator';
describe('trpc react generator', () => {
  let tree: Tree;
  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    // Mock frontend project configuration
    tree.write(
      'apps/frontend/project.json',
      JSON.stringify({
        name: 'frontend',
        root: 'apps/frontend',
        sourceRoot: 'apps/frontend/src',
      })
    );
    // Mock backend project configuration
    tree.write(
      'apps/backend/project.json',
      JSON.stringify({
        name: 'backend',
        root: 'apps/backend',
        sourceRoot: 'apps/backend/src',
        metadata: {
          apiName: 'TestApi',
        },
      })
    );
    // Mock main.tsx file
    tree.write(
      'apps/frontend/src/main.tsx',
      `
import { App } from './app';
import { BrowserRouter } from 'react-router-dom';

export function Main() {
  return <BrowserRouter><App /></BrowserRouter>;
}
`
    );
  });
  it('should generate trpc react files', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
      auth: 'None',
    });
    // Verify generated files
    expect(
      tree.exists('apps/frontend/src/components/TrpcClients/index.tsx')
    ).toBeTruthy();
    expect(tree.exists('apps/frontend/src/hooks/useTestApi.tsx')).toBeTruthy();
    // Create snapshots of generated files
    expect(
      tree.read('apps/frontend/src/hooks/useTestApi.tsx', 'utf-8')
    ).toMatchSnapshot('useTestApi.tsx');
    expect(
      tree.read('apps/frontend/src/components/TrpcClients/index.tsx', 'utf-8')
    ).toMatchSnapshot('TrpcClients-index.tsx');
    expect(
      tree.read(
        'apps/frontend/src/components/TrpcClients/IsolatedTrpcProvider.tsx',
        'utf-8'
      )
    ).toMatchSnapshot('TrpcClients-IsolatedTrpcProvider.tsx');
    expect(
      tree.read(
        'apps/frontend/src/components/TrpcClients/TrpcApis.tsx',
        'utf-8'
      )
    ).toMatchSnapshot('TrpcClients-TrpcApis.tsx');
    expect(
      tree.read(
        'apps/frontend/src/components/TrpcClients/TrpcClientProviders.tsx',
        'utf-8'
      )
    ).toMatchSnapshot('TrpcClients-TrpcClientProviders.tsx');
  });
  it('should modify main.tsx correctly', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
      auth: 'None',
    });
    const mainTsxContent = tree.read('apps/frontend/src/main.tsx', 'utf-8');
    // Create snapshot of modified main.tsx
    expect(mainTsxContent).toMatchSnapshot('main.tsx');
  });
  it('should add required dependencies', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
      auth: 'None',
    });
    const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));
    // Verify dependencies were added
    expect(packageJson.dependencies['@trpc/react-query']).toBeDefined();
    expect(packageJson.dependencies['@tanstack/react-query']).toBeDefined();
  });
  it('should handle IAM auth option', async () => {
    await reactGenerator(tree, {
      frontendProjectName: 'frontend',
      backendProjectName: 'backend',
      auth: 'IAM',
    });
    const trpcProviderContent = tree.read(
      'apps/frontend/src/components/TRPCClientProvider/index.tsx',
      'utf-8'
    );
    expect(trpcProviderContent).toMatchSnapshot('TRPCClientProvider-IAM.tsx');
  });
});
