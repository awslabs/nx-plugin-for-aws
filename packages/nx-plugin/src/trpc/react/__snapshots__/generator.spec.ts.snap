// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`trpc react generator > should generate trpc react files > TrpcClients-IsolatedTrpcProvider.tsx 1`] = `
"import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { createTRPCReact } from '@trpc/react-query';
import {
  httpBatchLink,
  httpLink,
  HTTPBatchLinkOptions,
  HTTPLinkOptions,
  splitLink,
} from '@trpc/client';
import { useState, FC, createContext, useMemo, PropsWithChildren } from 'react';

import { AnyTRPCRouter } from '@trpc/server';

export interface IsolatedTrpcClientProviderProps extends PropsWithChildren {
  readonly apiUrl: string;
}

export const createIsolatedTrpcClientProvider = <
  TAppRouter extends AnyTRPCRouter,
  TContext,
>() => {
  const isolatedTrpcContext = createContext(null as any);
  const trpc = createTRPCReact<TAppRouter, TContext>({
    context: isolatedTrpcContext,
  });

  const IsolatedTrpcClientProvider: FC<IsolatedTrpcClientProviderProps> = ({
    apiUrl,
    children,
  }) => {
    const [queryClient] = useState(() => new QueryClient());

    const trpcClient = useMemo(() => {
      const linkOptions: HTTPLinkOptions<any> & HTTPBatchLinkOptions<any> = {
        url: apiUrl,
      };

      // @ts-expect-error type errors due to unknown concrete AppRouter
      return trpc.createClient({
        links: [
          splitLink({
            condition(op) {
              return op.context.skipBatch === true;
            },
            true: httpLink(linkOptions),
            false: httpBatchLink(linkOptions),
          }),
        ],
      });
    }, [apiUrl, sigv4Client]);

    return (
      // @ts-expect-error type errors due to unknown concrete AppRouter
      <trpc.Provider client={trpcClient} queryClient={queryClient}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
        {/* @ts-expect-error type errors due to unknown concrete AppRouter */}
      </trpc.Provider>
    );
  };

  return {
    trpc,
    Provider: IsolatedTrpcClientProvider,
  };
};
"
`;

exports[`trpc react generator > should generate trpc react files > TrpcClients-TrpcApis.tsx 1`] = `
"import {
  AppRouter as TestApiAppRouter,
  Context as TestApiContext,
} from 'backend';
import { createIsolatedTrpcClientProvider } from './IsolatedTrpcProvider';
export default {
  TestApi: createIsolatedTrpcClientProvider<TestApiAppRouter, TestApiContext>(),
};
"
`;

exports[`trpc react generator > should generate trpc react files > TrpcClients-TrpcClientProviders.tsx 1`] = `
"import TrpcApis from './TrpcApis';
import { useRuntimeConfig } from '../../hooks/useRuntimeConfig';
import { FC, PropsWithChildren } from 'react';
const TrpcClientProviders: FC<PropsWithChildren> = ({ children }) => {
  const runtimeConfig = useRuntimeConfig();
  return (
    <TrpcApis.TestApi.Provider apiUrl={runtimeConfig.httpApis.TestApi}>
      {children}
    </TrpcApis.TestApi.Provider>
  );
};
export default TrpcClientProviders;
"
`;

exports[`trpc react generator > should generate trpc react files > TrpcClients-index.tsx 1`] = `
"import TrpcClientProviders from './TrpcClientProviders';
import _TrpcApis from './TrpcApis';

export const TrpcApis = _TrpcApis;
export default TrpcClientProviders;
"
`;

exports[`trpc react generator > should generate trpc react files > useTestApi.tsx 1`] = `
"import { TrpcApis } from '../components/TrpcClients';

export const useTestApi = () => TrpcApis.TestApi.trpc;
"
`;

exports[`trpc react generator > should handle IAM auth option > TRPCClientProvider-IAM.tsx 1`] = `null`;

exports[`trpc react generator > should modify main.tsx correctly > main.tsx 1`] = `
"import TrpcClientProviders from './components/TrpcClients';
import RuntimeConfigProvider from './components/RuntimeConfig';
import { App } from './app';
import { RouterProvider } from '@tanstack/react-router';
export function Main() {
  return (
    <RuntimeConfigProvider>
      <TrpcClientProviders>
        <RouterProvider router={router} />
      </TrpcClientProviders>
    </RuntimeConfigProvider>
  );
}
"
`;
